beets/dbcore/query.py

```diff
@@ -550,19 +551,38 @@ def parse(cls, string):
         """Parse a date and return a `Period` object or `None` if the
         string is empty.
         """
-        if not string:
-            return None
-        ordinal = string.count('-')
-        if ordinal >= len(cls.date_formats):
-            # Too many components.
-            return None
-        date_format = cls.date_formats[ordinal]
-        try:
-            date = datetime.strptime(string, date_format)
-        except ValueError:
-            # Parsing failed.
-            return None
-        precision = cls.precisions[ordinal]
+        pattern_dq = '@([+|-]?)(\d+)([y|m|w|d])'
+        if re.match(pattern_dq, string) is not None:
+            sign = re.match(pattern_dq, string).group(1)
+            quantity = re.match(pattern_dq, string).group(2)
+            timespan = re.match(pattern_dq, string).group(3)
```

Comment:
You can save the result of `re.match(pattern_dq, string)` in a variable and reuse it four times.